torii/src/utils.rs

```diff
@@ -37,48 +24,97 @@ impl<T: Encode + Send> Reply for Scale<T> {
     }
 }
 
-/// Adds state to filter
-macro_rules! add_state {
-    ( $( $state : expr ),* $(,)? ) => {
-        warp::any().map({
-            let state = ($( $state.clone(), )*);
-            move || state.clone()
-        }).untuple_one()
-    }
-}
-
-pub mod body {
-    use iroha_version::error::Error as VersionError;
+pub mod extractors {
+    use axum::{
+        async_trait,
+        body::Bytes,
+        extract::{FromRequest, FromRequestParts, Query, Request},
+        http::StatusCode,
+    };
+    use iroha_data_model::query::cursor::ForwardCursor;
 
     use super::*;
 
-    /// Decode body as versioned scale codec
-    pub fn versioned<T: DecodeVersioned>() -> impl Filter<Extract = (T,), Error = Rejection> + Copy
+    /// Extractor of scale encoded versioned data from body
+    #[derive(Clone, Copy, Debug)]
+    pub struct ScaleVersioned<T>(pub T);
+
+    #[async_trait]
+    impl<S, T> FromRequest<S> for ScaleVersioned<T>
+    where
+        Bytes: FromRequest<S>,
+        S: Send + Sync,
+        T: DecodeVersioned,
     {
-        warp::body::bytes().and_then(|body: Bytes| async move {
-            T::decode_all_versioned(body.as_ref()).map_err(warp::reject::custom)
-        })
+        type Rejection = Response;
+
+        async fn from_request(req: Request, state: &S) -> Result<Self, Self::Rejection> {
+            let body = Bytes::from_request(req, state)
+                .await
+                .map_err(IntoResponse::into_response)?;
+
+            T::decode_all_versioned(&body)
+                .map(ScaleVersioned)
+                .map_err(|err| {
+                    (
+                        axum::http::StatusCode::BAD_REQUEST,
+                        format!("Transaction Rejected (Malformed), Reason : '{err}'"),
```

Comment:
Not sure it should say "transaction" here. AFAIU, it's also used for queries too..

```suggestion
                        format!("Could not decode request: {err}"),
```